### PR TITLE
06 email

### PR DESCRIPTION
**Pull request by nNg9999**

### CODE REVIEW

File: src/email.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+// require('dotenv').config();
+
+const config = require('../config.js');
+const sgMail = require('@sendgrid/mail')
+
+// sgMail.setApiKey(process.env.SENDGRID_API_KEY)
+sgMail.setApiKey(config.sendGridApiKey)
+
+const msg = {
+  to: ['reloader.zhelinskyi@gmail.com', 'mykola.zhelinskyi@gmail.com'], // Change to your recipient
```
Comment (odykyi) [inline src/email.js]: move all emails to config

File: src/email.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  from: 'mykola.zhelinskyi@gmail.com', // Change to your verified sender
```
Comment (odykyi) [inline src/email.js]: move all emails to config

File: src/email.js [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  subject: 'Sending with SendGrid is Fun',
```
Comment (odykyi) [inline src/email.js]: move subject, text, html to config

File: src/email.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  text: 'and easy to do anywhere, even with Node.js',
+  html: '<strong>and easy to do anywhere, even with Node.js</strong>',
+}
+// sgMail
```
Comment (odykyi) [inline src/email.js]: remove commented code

File: src/helpers/create-jwt-token.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+const jwt = require("jsonwebtoken");
+const config = require("../../config");
+
+module.exports = (options = {}) => {
+  return jwt.sign("", config.jwtPrivateKey, options);
```
Comment (odykyi) [inline src/helpers/create-jwt-token.js:5]: do not use empty string 
you should put user id to jwt

File: src/helpers/create-jwt-token.js [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+};
+
+// const createJwt = (options = {}) => {
```
Comment (odykyi) [inline src/helpers/create-jwt-token.js]: remove commented code

File: src/routers/authRouter.js [hunk 7]
```diff
@@ -37,14 +40,72 @@ router.post("/register", errorWrapper(async (req, res) => {
   const createdUser = await UserModel.create({
     email,
     password,
-    avatarURL
+    avatarURL,
+    valid: false,
+    verificationToken: createJwtToken()
+    // otp: {
```
Comment (odykyi) [inline src/routers/authRouter.js]: remove commented code

File: src/routers/authRouter.js [hunk 8]
```diff
@@ -37,14 +40,72 @@ router.post("/register", errorWrapper(async (req, res) => {
...
 
 }));
 
 
+router.get("/verify/:verificationToken", errorWrapper(async (req, res) => {
+
+  const { verificationToken } = req.params;
+  // console.log(otp);
+
+  // const user = await UserModel.findOne({
```
Comment (odykyi) [inline src/routers/authRouter.js]: remove commented

File: src/routers/authRouter.js [hunk 9]
```diff
@@ -37,14 +40,72 @@ router.post("/register", errorWrapper(async (req, res) => {
...
+  // });
+
+  const user = await UserModel.findOne({
+    verificationToken
+  });
+
+  if (!user) {
+    logger.error("User not found");
+    throw new ApiError(404, "User not found");
+    // return res.redirect("/app/error?type=user-not-found");
```
Comment (odykyi) [inline src/routers/authRouter.js]: remove commented

File: src/routers/authRouter.js [hunk 10]
```diff
@@ -37,14 +40,72 @@ router.post("/register", errorWrapper(async (req, res) => {
...
+    // return res.redirect("/error.html");
+  }
+
+  user.valid = true
+  user.verificationToken = null;
+  await user.save();
+
+  // if (user.valid) {
```
Comment (odykyi) [inline src/routers/authRouter.js]: remove commented

File: src/services/mailer.js [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+const { getLogger } = require("../helpers");
+const config = require("../../config");
+
+class Mailer {
+  constructor() {
+    this.logger = getLogger("mailer");
+
+    this.transporter = nodemailer.createTransport(
+      {
+        host: "smtp.sendgrid.net",
```
Comment (odykyi) [inline src/services/mailer.js]: move to config

File: src/services/mailer.js [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        port: 587,
```
Comment (odykyi) [inline src/services/mailer.js]: move to config

File: src/services/mailer.js [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+        secure: false, // true for 465, false for other ports
+        auth: {
+          user: 'apikey', // generated ethereal user
```
Comment (odykyi) [inline src/services/mailer.js]: move to config

File: src/services/mailer.js [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+class Mailer {
+  constructor() {
+    this.logger = getLogger("mailer");
+
+    this.transporter = nodemailer.createTransport(
+      {
+        host: "smtp.sendgrid.net",
+        port: 587,
+        secure: false, // true for 465, false for other ports
```
Comment (odykyi) [inline src/services/mailer.js]: move to config

File: src/services/mailerWebApi.js [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+const { getLogger } = require("../helpers");
+const config = require("../../config");
+
+class MailerWebApi {
+  constructor() {
+    this.logger = getLogger("mailerWebApi");
+
+    this.transporter = nodemailer.createTransport(
+      {
+        service: 'gmail', // true for 465, false for other ports
```
Comment (odykyi) [inline src/services/mailerWebApi.js]: move to config

File: src/services/mailerWebApi.js [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+  async sendBody({ to, subject, body }) {
+    return this.sendMailFromSupport({ to, subject, html: body });
+  }
+}
+
+
+module.exports = new MailerWebApi();
+
+// mailerWebApi = new MailerWebApi();
```
Comment (odykyi) [inline src/services/mailerWebApi.js]: remove commented

File: src/socket-handlers/create-char-handler.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
+const BaseHandler = require("./base-handler");
+const { ChatModel } = require("../database/modles");
```
Comment (odykyi) [inline src/socket-handlers/create-char-handler.js]: looks like ChatModel  not exists https://github.com/nNg9999/node-js-30/tree/06-email/src/database/models

File: src/socket-handlers/send-message-handler.js [hunk 18]
```diff
@@ -0,0 +1,31 @@
+const BaseHandler = require("./base-handler");
+const { ChatModel, MessageModel } = require("../database/modles");
```
Comment (odykyi) [inline src/socket-handlers/send-message-handler.js]: ChatModel not exists

File: src/services/mailer.js [hunk 19]
```diff
@@ -0,0 +1,54 @@
...
+
+  async sendMailFromSupport({ to, subject, text, html }) {
+    this.logger.debug("Sending email");
+    this.logger.debug("to ->", to);
+    this.logger.debug("subject ->", subject);
+    text && this.logger.debug("text ->", text);
+    html && this.logger.debug("html ->", html);
+
+    return this.transporter.sendMail({
+      from: '"Support 👻" <mykola.zhelinskyi@gmail.com>', // sender address
```
Comment (odykyi) [inline src/services/mailer.js:26]: move sender address to config

**Review by odykyi — CHANGES_REQUESTED** [review]
comments

**Comment by odykyi** [comment]
@nNg9999 wait for changes

File: src/email.js (hunk 1 above)
Comment (nNg9999) [reply src/email.js]: Move

File: src/email.js (hunk 2 above)
Comment (nNg9999) [reply src/email.js]: Move

File: src/email.js (hunk 3 above)
Comment (nNg9999) [reply src/email.js]: Move

File: src/email.js (hunk 4 above)
Comment (nNg9999) [reply src/email.js]: remove

File: src/helpers/create-jwt-token.js (hunk 5 above)
Comment (nNg9999) [reply src/helpers/create-jwt-token.js:5]: this function generates only a token it does not need a user ID

File: src/helpers/create-jwt-token.js (hunk 6 above)
Comment (nNg9999) [reply src/helpers/create-jwt-token.js]: remove

File: src/routers/authRouter.js (hunk 7 above)
Comment (nNg9999) [reply src/routers/authRouter.js]: remove

File: src/routers/authRouter.js (hunk 8 above)
Comment (nNg9999) [reply src/routers/authRouter.js]: remove

File: src/routers/authRouter.js (hunk 9 above)
Comment (nNg9999) [reply src/routers/authRouter.js]: remove

File: src/routers/authRouter.js (hunk 10 above)
Comment (nNg9999) [reply src/routers/authRouter.js]: remove

File: src/services/mailer.js (hunk 11 above)
Comment (nNg9999) [reply src/services/mailer.js]: move

File: src/services/mailer.js (hunk 12 above)
Comment (nNg9999) [reply src/services/mailer.js]: move

File: src/services/mailer.js (hunk 13 above)
Comment (nNg9999) [reply src/services/mailer.js]: move

File: src/services/mailer.js (hunk 14 above)
Comment (nNg9999) [reply src/services/mailer.js]: move

File: src/services/mailerWebApi.js (hunk 15 above)
Comment (nNg9999) [reply src/services/mailerWebApi.js]: move

File: src/services/mailerWebApi.js (hunk 16 above)
Comment (nNg9999) [reply src/services/mailerWebApi.js]: remove

File: src/socket-handlers/create-char-handler.js (hunk 17 above)
Comment (nNg9999) [reply src/socket-handlers/create-char-handler.js]: add ChatModel

File: src/socket-handlers/send-message-handler.js (hunk 18 above)
Comment (nNg9999) [reply src/socket-handlers/send-message-handler.js]: add ChatModel

File: src/services/mailer.js (hunk 19 above)
Comment (nNg9999) [reply src/services/mailer.js:26]: move

**Review by odykyi — APPROVED** [review]

**Review by odykyi — APPROVED** [review]

**Pull request merged**